### PR TITLE
Add wildcard search in manage_user_page.php

### PR DESCRIPTION
**Pull request by CasN**
Added wildcard search on realname including the option to exclude a pattern.
So searchterm "jan" would lead to wildcard "%JAN%"
but searchterm "jan -dek" would lead to include  "%JAN%" and to exclude "%DEK%"

Fixes [#12677](https://mantisbt.org/bugs/view.php?id=12677)

### CODE REVIEW

File: manage_user_page.php [hunk 1]
```diff
@@ -185,6 +190,17 @@
 $t_where_params = array();
 if( $f_filter === 'ALL' ) {
 	$t_where = '(1 = 1)';
+} else if( $f_findname<>"" ) {
+	$c_prefix = strtoupper(db_prepare_string($f_filter));
```
Comment (atrol) [inline manage_user_page.php]: function db_prepare_string is deprecated and should not be used any longer, see https://github.com/mantisbt/mantisbt/blob/release-2.16.0/core/database_api.php#L620

File: manage_user_page.php [hunk 2]
```diff
@@ -185,6 +190,17 @@
...
+	$pos = strpos($c_prefix, '-');
+	if ($pos){
+		$exclude = trim(substr($c_prefix,$pos+1));
+		$c_prefix= trim(substr($c_prefix, 0,$pos-1));
+		$t_where = "(upper(realname) like '%$c_prefix%' and upper(realname) not like '%$exclude%')";
```
Comment (atrol) [inline manage_user_page.php]: There is a function db_helper_like that provides case-insensitive like phrase based on the current db type (e.g. works with MySQL and PostgreSQL, see https://github.com/mantisbt/mantisbt/blob/release-2.16.0/core/database_api.php#L726
I expect a better performance on installations with lot's of users as I assume that your implementation leads to a full table scan.

File: manage_user_page.php [hunk 3]
```diff
@@ -185,6 +190,17 @@
...
+	} else{
+		$c_prefix= trim(substr($c_prefix, 0,$pos-1));
+		$t_where = "(upper(realname) like '%$c_prefix%')";
```
Comment (atrol) [inline manage_user_page.php]: Same as above

**Comment by atrol** [comment]
- There should be no difference of the buttons in terms of styling
managuser

- Bug when entering a single character in search field
You don't get the expected users, e.g. enter `a` but get a list of users without `a` in real name or empty real name

**Comment by CasN** [comment]
Roland, I have adressed all your issues and fixed the bug. In additiojn I have added a function "db_helper_not_like" to the database_api to bring all inline.

**Comment by cproensa** [comment]
> I have added a function "db_helper_not_like" to the database_api to bring all inline.

What is the difference between `A not like 'E'` and `not A like 'E'` ?
If none, using the second construct there is no need for the new function

**Comment by dregad** [comment]
> Please be more specific 

http://www.mantisbt.org/wiki/doku.php/mantisbt:coding_guidelines#braces_and_parentheses
Quite a few space around parenthesis. Also add spaces around the operator in expressions.

File: manage_user_page.php [hunk 4]
```diff
@@ -64,6 +64,11 @@
 $f_filter        = mb_strtoupper( gpc_get_string( 'filter', config_get( 'default_manage_user_prefix' ) ) );
 $f_page_number   = gpc_get_int( 'page_number', 1 );
 
+$f_findname = gpc_get_string( 'findname', '' ); 
+if($f_findname<>""){
```
Comment (dregad) [inline manage_user_page.php]: here

File: manage_user_page.php [hunk 5]
```diff
@@ -64,6 +64,11 @@
...
+	$f_filter=$f_findname;
```
Comment (dregad) [inline manage_user_page.php]: here

File: manage_user_page.php [hunk 6]
```diff
@@ -185,6 +190,21 @@
...
+	$c_prefix = strtoupper($f_filter);
+	$t_pos = strpos($c_prefix, '-');
+	if($t_pos){
+		$t_exclude = trim(substr($c_prefix,$t_pos+1));
+		$c_prefix= trim(substr($c_prefix, 0,$t_pos-1));
+		$t_where_params[] ='%'.$c_prefix . '%';
+		$t_where = db_helper_like( 'UPPER(realname)' );
+		$t_where .= " AND NOT ";
+		$t_where_params[] ='%'.$t_exclude . '%';
+		$t_where .= db_helper_like( 'UPPER(realname)' );
```
Comment (dregad) [inline manage_user_page.php]: here and the above 10 lines

File: manage_user_page.php [hunk 7]
```diff
@@ -64,6 +64,8 @@
 $f_filter        = mb_strtoupper( gpc_get_string( 'filter', config_get( 'default_manage_user_prefix' ) ) );
 $f_page_number   = gpc_get_int( 'page_number', 1 );
 
+$f_findname	 = mb_strtoupper( gpc_get_string( 'findname', '' ) );
```
Comment (atrol) [inline manage_user_page.php]: mb_strtoupper not needed as db_helper_like provides case-insensitive comparison.

File: manage_user_page.php [hunk 8]
```diff
@@ -185,6 +187,19 @@
 $t_where_params = array();
 if( $f_filter === 'ALL' ) {
 	$t_where = '(1 = 1)';
+} else if( $f_findname <> "" ) {
+	if( $t_pos ){
+		$t_exclude		= trim( substr( $f_findname,$t_pos+1 ) );
+		$f_findname		= trim( substr( $f_findname, 0,$t_pos-1 ) );
+		$t_where_params[]	= '%'.$f_findname . '%';
+		$t_where		= db_helper_like( 'UPPER(realname)' );
```
Comment (atrol) [inline manage_user_page.php]: UPPER not needed as db_helper_like provides case-insensitive comparison.

File: manage_user_page.php [hunk 9]
```diff
@@ -185,6 +187,19 @@
...
+		$t_where		.= " AND NOT ";
+		$t_where_params[] 	= '%'.$t_exclude . '%';
+		$t_where 		.= db_helper_like( 'UPPER(realname)' );
```
Comment (atrol) [inline manage_user_page.php]: same as above

File: manage_user_page.php [hunk 10]
```diff
@@ -185,6 +187,19 @@
...
+	} else {
+		$t_where_params[] 	= '%'.$f_findname . '%';
+		$t_where 		= db_helper_like( 'UPPER(realname)' );
```
Comment (atrol) [inline manage_user_page.php]: same as above

**Comment by atrol** [comment]
The functionality is broken after the latest change.

**Comment by atrol** [comment]
After searching for a name and getting the result list, the UI does not work as expected.
E.g. the filter is lost when sorting columns, clicking the `Apply filter` button or clicking the `Next` button.

**Comment by atrol** [comment]
@CasN using the master branch for PR's is no good idea in general.
Using the same branch for different PR's is even worse, as 
E.g. this PR contains now also the changes of your PR https://github.com/mantisbt/mantisbt/pull/1380
You should create a separate branches.
https://www.mantisbt.org/docs/master/en-US/Developers_Guide/html-desktop/#dev.contrib.prepare
https://www.mantisbt.org/docs/master/en-US/Developers_Guide/html-desktop/#dev.contrib.submit.github

**Comment by atrol** [comment]
@cas let me know if you want to close this one in favor of https://github.com/mantisbt/mantisbt/pull/1381 where I

- rebased based on latest master
- squashed commits while keeping commits of cas and dregad
- minor code style corrections
- removed unrelated changes (config_defaults_inc.php)
- replaced `User Name` label by placeholder `Username, Real Name or E-mail`

**Comment by atrol** [comment]
Closed in favor of https://github.com/mantisbt/mantisbt/pull/1381

**Pull request closed without merging**